WebPage_LoadRequest IPC fails decoding in PingDuoDuo app

cdumez opened the pull request:
#### 0ba9e6c4473fcc7d78aaa785794cbd884e61a094
<pre>
WebPage_LoadRequest IPC fails decoding in PingDuoDuo app
<a href="https://bugs.webkit.org/show_bug.cgi?id=258486">https://bugs.webkit.org/show_bug.cgi?id=258486</a>
rdar://111161160

Reviewed by Tim Horton.

WebPage_LoadRequest IPC was failing decoding in PingDuoDuo app.
The issue was due to getting a WKSecureCodingURLWrapper instead
of a NSURL when decoding the baseURL of a NSURL.

I am not sure how we ended up in this situation but I made the
bug go away by simplifying the code. The coder used to encode
the URL in two parts:
1. The baseURL
2. The bytes from the URL&apos;s relative string

Then it would decode the URL in 2 parts:
1. The baseURL
2. The bytes from the URL&apos;s relative string

It would then call CFURLCreateAbsoluteURLWithBytes() with those
2 parts, which would result in an *absolute* URL. The information
about baseURL / relative string would be lost.

As a result, I have decided to simply encode the URL in one part,
the absolute URL bytes. The decoding results ends up being the
same (an absolute URL). It simplifies both coding and decoding
and makes the bug go away since it was about decoding baseURLs.

* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(-[WKSecureCodingURLWrapper encodeWithCoder:]):
(-[WKSecureCodingURLWrapper initWithCoder:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/IPCTestingAPI.mm:

Canonical link: <a href="https://commits.webkit.org/265529@main">https://commits.webkit.org/265529@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/477aef53592f781b8419ee83936d80bf1abc35b8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11013 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11225 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11558 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12662 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10511 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11028 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13602 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11207 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13445 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11172 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12069 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9292 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13067 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9362 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9967 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/17171 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10433 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10121 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13340 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10542 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/8628 "35 flakes 23 failures") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9714 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2668 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13985 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10396 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham commented on Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
```diff
@@ -191,16 +184,9 @@ - (instancetype)initWithCoder:(NSCoder *)coder
     NSUInteger length;
     if (auto bytes = (UInt8 *)[coder decodeBytesWithReturnedLength:&length]) {
-        m_wrappedURL = bridge_cast(adoptCF(CFURLCreateAbsoluteURLWithBytes(nullptr, bytes, length, kCFStringEncodingUTF8, bridge_cast(baseURL.get()), true)));
+        m_wrappedURL = bridge_cast(adoptCF(CFURLCreateAbsoluteURLWithBytes(nullptr, bytes, length, kCFStringEncodingUTF8, nullptr, true)));
```
Could it be that this is wrong? Maybe it needs to be using the baseURL to create the true URL?

brentfulgham replied:
I looked through the CFNetwork API, and don't see a Create function that takes baseURL (maybe I'm missing something).

In that case, the special handling of baseURL is silly, and maybe we can do away with the `WKSecureCodingURLWrapper` class entirely?

brentfulgham replied:
Oh, wait! I'm wrong! This function does accept baseURL:

[CFURLRef](https://developer.apple.com/documentation/corefoundation/cfurlref?language=objc)
```
CFURLCreateAbsoluteURLWithBytes(CFAllocatorRef alloc, const UInt8 *relativeURLBytes, CFIndex length, CFStringEncoding encoding, CFURLRef baseURL, Boolean useCompatibilityMode);
```

brentfulgham replied:
We should be doing a bridge-cast of the baseURL to a CFURLRef to add to the m_wrappedURL creation.

cdumez replied:
Why do we want to do away with WKSecureCodingURLWrapper? I thought this was a convenient way to code/decode a URL as bytes.

cdumez replied:
> Could it be that this is wrong? Maybe it needs to be using the baseURL to create the true URL?

We used to pass a baseURL because the `bytes` only contained the relative part of the URL. We needed both the baseURL + related part bytes to construct the absolute URL.

Now though, I encode the absolute URL as the bytes. As a result, we don't need to pass a base URL. Note that this is covered by my new API test.

brentfulgham replied:
Ah, very nice.,

cdumez committed:
WebPage_LoadRequest IPC fails decoding in PingDuoDuo app
https://bugs.webkit.org/show_bug.cgi?id=258486
rdar://111161160

Reviewed by Tim Horton.

WebPage_LoadRequest IPC was failing decoding in PingDuoDuo app.
The issue was due to getting a WKSecureCodingURLWrapper instead
of a NSURL when decoding the baseURL of a NSURL.

I am not sure how we ended up in this situation but I made the
bug go away by simplifying the code. The coder used to encode
the URL in two parts:
1. The baseURL
2. The bytes from the URL's relative string

Then it would decode the URL in 2 parts:
1. The baseURL
2. The bytes from the URL's relative string

It would then call CFURLCreateAbsoluteURLWithBytes() with those
2 parts, which would result in an *absolute* URL. The information
about baseURL / relative string would be lost.

As a result, I have decided to simply encode the URL in one part,
the absolute URL bytes. The decoding results ends up being the
same (an absolute URL). It simplifies both coding and decoding
and makes the bug go away since it was about decoding baseURLs.

* Source/WebKit/Shared/Cocoa/ArgumentCodersCocoa.mm:
(-[WKSecureCodingURLWrapper encodeWithCoder:]):
(-[WKSecureCodingURLWrapper initWithCoder:]):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/IPCTestingAPI.mm:

Canonical link: https://commits.webkit.org/265529@main
